src/java.base/share/classes/sun/security/provider/ML_DSA.java

```diff
@@ -0,0 +1,1229 @@
+            throw new SignatureException("Incorrect signature length");
+        }
+
+        //Decode cTilde
+        byte[] cTilde = Arrays.copyOfRange(sig, 0, lambda/4);
+
+        //Decode z
+        int start = cSize;
+        int end = start + zSize;
+        int[][] z = new int[mlDsa_l][ML_DSA_N];
+        bitUnpack(z, sig, start, mlDsa_l, gamma1, gamma1Bits + 1);
+
+        //Decode h
+        start = end;
+        boolean[][] h = hintBitUnpack(sig, start);
+        if (h == null) {
+            throw new SignatureException("Invalid hints encoding");
+        }
+
+        return new ML_DSA_Signature(cTilde, z, h);
+    }
+
+    /*
+    Auxiliary functions defined in Section 7.3 of specification
+     */
+
+    private class Shake256Slicer {
+        SHAKE256 xof;
+        byte[] block;
+        int byteOffset;
+        int current;
+        int bitsInCurrent;
+        int bitsPerCall;
+        int bitMask;
+
+        Shake256Slicer(SHAKE256 xof, int bitsPerCall) {
+            this.xof = xof;
+            //BitsPerCall can only be 4 (when called from sampleS1S2),
+            //or 8 (when called from sampleInBall)
+            this.bitsPerCall = bitsPerCall;
+            bitMask = (1 << bitsPerCall) - 1;
+            current = 0;
+            byteOffset = SHAKE256_BLOCK_SIZE;
+            bitsInCurrent = 0;
+            block = new byte[SHAKE256_BLOCK_SIZE];
+        }
+
+        void reset() {
+            xof.reset();
+            current = 0;
+            byteOffset = SHAKE256_BLOCK_SIZE;
+            bitsInCurrent = 0;
+        }
+
+        int squeezeBits() {
+            while (bitsInCurrent < bitsPerCall) {
+                if (byteOffset == SHAKE256_BLOCK_SIZE) {
+                    xof.squeeze(block, 0, SHAKE256_BLOCK_SIZE);
+                    byteOffset = 0;
+                }
+                current += ((block[byteOffset++] & 0xff) << bitsInCurrent);
+                bitsInCurrent += 8;
+            }
+            int result = current & bitMask;
+            current >>= bitsPerCall;
+            bitsInCurrent -= bitsPerCall;
+            return result;
+        }
+    }
+
+    private void sampleInBall(int[] c, byte[] rho) {
+        var xof = new SHAKE256(0);
+        Shake256Slicer slicer = new Shake256Slicer(xof, 8);
+        xof.update(rho);
+
+        long parity = 0;
+        for (int i = 0; i < 8; i++) {
+            long sample = slicer.squeezeBits();
+            parity |= sample << 8 * i;
+        }
+
+        Arrays.fill(c, 0);
+
+        int k = 8;
+        for (int i = 256 - tau; i < 256; i++) {
+            //Get random index < i
+            int j = slicer.squeezeBits();
+            while (j > i) {
+                j = slicer.squeezeBits();
+            }
+
+            //Swap c[i] and c[j], set c[j] based on parity
+            c[i] = c[j];
+            c[j] = (int) (1 - 2 * (parity & 1));
+            parity >>= 1;
+        }
+    }
+
+    int[][][] generateA(byte[] seed) {
```

Comment:
I noticed in your generateA method for ML-KEM, that you allocate your 3-dim array as ` short[][][] a = new short[mlKem_k][mlKem_k][];`  Since you're going to effectively return all those 3rd dimension arrays you initially allocated back to the GC once you've assigned them all down on line 869, could you do the same thing here?